### ISSUE TITLE
[MNG-1388] Transitive Dependencies in a profile are not used

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#3541]
**[Damian Bradicich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dbradicich)** opened **[MNG-1388](https://issues.apache.org/jira/browse/MNG-1388?redirect=false)** and commented

I have a jar project file that defines a dependency inside a certain profile.  If I then include that project inside of another war project, the dependencies defined in the jar project's profile isn't getting transferred over to the war.

Ie we have this:
A depends on SQL or Oracle depending on profile
B depends on A.
If sql profile is active, I would expect that when I build B, it pulls
the transitive dependancy on sql from A.

---

**Affects:** 2.0

**Attachments:**
- [mng1388.pom](https://issues.apache.org/jira/secure/attachment/12961442/mng1388.pom) (_336 bytes_)

**Issue Links:**
- [MNG-4985](https://issues.apache.org/jira/browse/MNG-4985) Transitive dependencies with classifiers not resolved correctly
 (_**"is duplicated by"**_)
- [MNG-4894](https://issues.apache.org/jira/browse/MNG-4894) Profile is ignored for dependencies
 (_**"is duplicated by"**_)
- [MNG-5691](https://issues.apache.org/jira/browse/MNG-5691) Reopen MNG-1388
 (_**"is duplicated by"**_)
- [MNG-4299](https://issues.apache.org/jira/browse/MNG-4299) dependency profile activation

- [MNG-4531](https://issues.apache.org/jira/browse/MNG-4531) Add option to use a profile of a dependency

- [MNG-3034](https://issues.apache.org/jira/browse/MNG-3034) Import a profile into a POM

- [MNG-4538](https://issues.apache.org/jira/browse/MNG-4538) dependencyManagament dependencies within profiles are not activated by settings.xml

26 votes, 30 watchers

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

are you able to produce a small test case to demonstrate this?

**Comment by jira-importer** [comment]
**[Damian Bradicich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dbradicich)** commented

Here is an example of 2 pom files that will show the problem, also note that the active profile is defined in settings.xml

\<project>
\<modelVersion>4.0.0\</modelVersion>
\<groupId>test\</groupId>
\<artifactId>a\</artifactId>
\<version>1.0\</version>
\<packaging>jar\</packaging>
\<name>a test\</name>
\<profiles>
\<profile>
\<id>struts-old\</id>
\<dependencies>
\<dependency>
\<groupId>struts\</groupId>
\<artifactId>struts\</artifactId>
\<version>1.1\</version>
\<scope>compile\</scope>
\</dependency>
\</dependencies>
\</profile>
\<profile>
\<id>struts-new\</id>
\<dependencies>
\<dependency>
\<groupId>struts\</groupId>
\<artifactId>struts\</artifactId>
\<version>1.2.7\</version>
\<scope>compile\</scope>
\</dependency>
\</dependencies>  
\</profile>
\</profiles>
\</project>
\<project>
\<modelVersion>4.0.0\</modelVersion>
\<groupId>test\</groupId>
\<artifactId>b\</artifactId>
\<version>1.0\</version>
\<packaging>war\</packaging>
\<name>b test\</name>
\<dependencies>
\<dependency>
\<groupId>test\</groupId>
\<artifactId>a\</artifactId>
\<version>1.0\</version>
\<scope>compile\</scope>
\</dependency>
\</dependencies>   
\</project>

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

This happens because dependency POMs aren't currently candidates for profile triggering.

However, it seems we will have to trigger profiles for these POMs...otherwise, we'll lose the ability to target a build for a given platform where it will require the inclusion of a helper API.

**Comment by jira-importer** [comment]
**[Carlos Sanchez Gonzalez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlos)** commented

Set a better title

**Comment by jira-importer** [comment]
**[Milos Kleint](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkleint)** commented

since profiles are identified by ids, it will probably happen that these IDs will clash in completely unrelated 3rd party libraries. maybe the scope of dependency pom profile triggering should be limuted to the same groupId?

**Comment by jira-importer** [comment]
**[Rob ten Hove](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=robth)** commented

I would not like limiting the scope of the dependency pom profile triggering to the same groupId. I'm working on a very large project with a lot of groupId's. The complete project can run on a middleware project or a stub of it. To accomplish this, the data objects that the applications use to communicate with the middleware will be either the 'real' implementations or the stubbed ones. So for each data object we have two versions: "impl" and "stub". To get the right dependencies, each project that uses the data objects will have two profiles, called "impl" and "stub". Now if I want to build an application with Maven, I will activate either profile. I just discovered that this does not work with Maven right now!

So I have this situation regarding dependencies:

```
A
  - B
  - profile id="stub"
    - DO1_stub
  - profile id="impl"
    - DO1_impl

B
  - profile id="stub"
    - DO2_stub
  - profile id="impl"
    - DO2_impl
```

If I run "mvn -Pstub clean package" on A's pom, then I will have DO1_stub but **not** DO2_stub!

I think that the profile id problem should be solved by good naming conventions, not by limiting the scope since that will reduce the applicability a lot. For example if it is a local profile, prefix it with the artifactId. If it is to be used within a group, prefix it with the groupId etc.

**Comment by jira-importer** [comment]
**[Rob ten Hove](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=robth)** commented

Update: the situation described above does work when profile activation using properties is used:

```
<profile>
  <id>stub</id>
  <activation>
    <property>
      <name>stub</name>
    </property>
  </activation>
  ...
</profile>
```

Running "mvn -Pstub -Dstub clean package" on A's pom gives both DO1_stub **and** DO2_stub.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

I second Milos' concern about clashing profile ids. I think the approach outlined by Paul in MNG-4531 is more appropriate.

**Comment by jira-importer** [comment]
**[Ivan Bondarenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=buzzeri)** commented

We are also waiting for this bug solve (have issue similar to described in MNG-4985).
We have a lot of modules, which form a hierarchy (3 or 4 levels). Modules can be built with classifiers ("release", "debug" etc). When war project references these modules, transitive dependencies without any classifiers are used.
In fact that makes "classifier paradigm" useless.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Happy birthday!

8 years old and counting.

**Comment by jira-importer** [comment]
**[Damian Bradicich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dbradicich)** commented

yes, happy birthday my old friend ;)

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Turning on a profile id transitively at this point after not doing it for so long is bound to cause issues.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Jason,

Please re-open this issue. This makes it impossible to package native libraries using Maven. Native libraries have multiple classifiers (one per platform) which are activated by choosing a profile. Native libraries that depend on other native libraries need to depend on an artifact with a classifier, but it is impossible to do so because Maven disregard these transitive dependencies.

In short: I consider this issue a showstopper with no workaround.

**Comment by jira-importer** [comment]
**[Iain Coulter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iaincoulter)** commented

Want to also add a request to reopen this issue,  we have to use various workarounds to pull the correct native dependencies without a solution to the above problem.  In some cases we upload a different pom to that used during compile, in other cases we end up downloading all dependencies across every platform even though we may only need say windows-x86-32 components.

**Comment by jira-importer** [comment]
**[Wiebke Timm](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=veroni)** commented

Requesting to reopen this issue as well. We are using maven for native packages and need to workaround this issue. We have classifiers for the different platforms (win32-x86 and linux-gcc) and common packages/headers (classifier "header").

Please fix this!

**Comment by jira-importer** [comment]
**[Iain Coulter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iaincoulter)** commented

Using properties for the classifiers is not a complete solution if your building a multi platorm relase.
ie selecting both win32-x86 and linux-gcc (from Wiebkes post) using profiles for each then the property for the classifier in first profile will be overwritten by the property in the second!

**Comment by jira-importer** [comment]
**[Wiebke Timm](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=veroni)** commented

Not sure what you mean by "set those in the profiles". If you set the property in some parent, you would need a different parent (for that platform's profile) for different platforms. But that doesn't solve cases with managed dependencies:
For managed dependencies, a whole subtree is built with a certain platform (i.e. classifier) that's different from the platform of the rest of the tree, but still must be unique within that subtree. This cannot be addressed properly without modeling of transitive dependencies.

In any case, we have tried using properties and failed. We are using a workaround with environment variables and that still can't properly deal with the subtree problem described above.

**Comment by jira-importer** [comment]
**[Joerg Schaible](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joehni)** commented

> selecting both win32-x86 and linux-gcc (from Wiebkes post)

Where was there a requirement that both platform profiles have to be active **at the same time** ?

Let's make an example:

parent.xml:

```xml
...
<dependencyManagement>
 <dependencies>
  <dependency>
   <groupId>com.company.project</groupId>
   <artifactId>component</artifactId>
   <version>1.0</version>
   <classifier>${company.platform}</classifier>
  </dependency>
  ...
 </dependencies>
</dependencyManagement>
...
<profiles>
 <profile>
  <activation>
   <property>linux</property>
  </activation>
  <properties>
   <company.platform>linux-gcc</company.platform>
  </properties>
 </profile>
 ..
</profiles>
<properties>
 <company.platform>win32-x86</company.platform><!-- default -->
 ...
</properties>
```

parent.xml:

```xml
...
<dependencies>
 <dependency>
  <groupId>com.company.project</groupId>
  <artifactId>component</artifactId>
  <classifier>${company.platform}</classifier>
 </dependency>
 ...
</dependencies>
<properties>
 <!-- you can overwrite directly if no other platform is required for this project
 <company.platform>win32-x86</company.platform>
 -->
</properties>
```

In this example I've used the property "linux" as activator, but you might use activation based on OS is appropriate. Anyway, with this setup you can call Maven either with:

mvn -Dlinux clean install

or define this property in the settings.xml. You are even able to test a build for a new platform directly:

mvn -Dcompany.platform=cygwin-gcc clean install

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Defining properties fails to solve the problem because transitive dependencies can and do span across multiple organizations. Meaning,

Company A publishes libraries A1, A2, A3 which depend on libraries B1, B2, B3 written by company B, which depend on libraries C1, C2, C3 written by company C. Defining properties this way leaks information from the innermost dependency to the outermost one, meaning:

1. Anyone building a native library needs to also know how to build all their transitive dependencies (libraries written by other people). Extrapolate this across a couple more level and this gets very complicated very fast.
2. This approach requires property names to be unique across the entire build tree, but again we don't own all transitive dependencies (this spans multiple organizations).

This issue should be reopened.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Filed MNG-5691 asking for this issue to be reopened. I am counting on you guys to keep pushing on this until it is fixed!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Reopening by request. If someone still wan't this as `Won't Fix`, please provide an explanation for.

**Comment by jira-importer** [comment]
**[Iain Coulter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iaincoulter)** commented

I have taken a stab and seeing what is involved in this.  My plan would add an optional "profiles" element as part of the dependency, While resolving depenencies if this "profiles" entry is present then the maven profiles that match those ids are activated and its dependencies are added to the list of dependencies to be resolved (these of course could also contain a "profiles" entry.

\<dependency>
\<groupId>com.company.project\</groupId>
\<artifactId>component\</artifactId>
\<version>1.0\</version>
\<profiles>ProfileA,ProfileB\</profiles>  \
\</dependency>

To achieve this with a proper fix consists of two parts:

1) Update Maven xml parsing to handle the above "profiles"  (this is easy enough but does cause a change to the maven pom xsd)

2) Update Eclipse Aether to add support for maven like profiles and its dependencies

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

Regarding the first two example poms provided. This should not be supported because it would mess things up when deployed to a repository. The poms in the repository need to be self-contained. Having a requirement on the activation of some profile makes it unusable (was it -Pstruts-old or -Pstruts-new or both?) Makes no sense to deploy things like this anywhere.

**Comment by jira-importer** [comment]
**[Abhinay Agarwal](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abhinay)** commented

I am trying to upgrade a library to work with JDK 11. There are some dependencies which are no longer a part of the JDK, so I introduced a profile which would activate depending on the JDK and would take care of the missing dependencies:

```java
<profile>
    <id>JDK 11</id>
    <activation>
        <activeByDefault>false</activeByDefault>
        <jdk>11</jdk>
    </activation>
    <properties>
        <javafx.version>11-ea+24</javafx.version>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-controls</artifactId>
            <version>${javafx.version}</version>
        </dependency>
        <dependency>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-fxml</artifactId>
            <version>${javafx.version}</version>
        </dependency>
    </dependencies>
</profile>
```

The profile works well when I try to compile and install the library with JDK 11.

But when I try to use the library as a dependency in another application, maven is unable to resolve the transitive dependencies inside the profile. Is there a workaround for this problem?

**Comment by jira-importer** [comment]
**[Igor Tykhyy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=divstar)** commented

I'd also like to see this issue fixed (though I am not sure I could fix it myself). I'd like profiles to apply to **not only dependencies**, but also **child-modules** (specified under `<modules>`).

```java
A
    - B (child-module)
    - C (child-module)
```

Doing

```java
mvn clean install -Psome-profile
```

should apply the profile `some-profile` on **A** as well as on **B** and **C**, running the profile in each module containing it.

At least for me it does not work - especially not if modules are stored in profiles themselves (I only need some modules to be built in order to deploy them; aggregator-project-packaging is set to _pom_).

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[divStar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=divStar) download attached pom-file (optionally rename it)
`mvn -f mng1388.pom verify` succeeds building only A
`mvn -f mng1388.pom verify -Psome-profile` fails due to missing directories B and C (which makes sense).
Profiled modules are recognized, so I don't the issue.
